File: 2015/02/node/main.mjs
import { readFile } from "node:fs/promises"
import { join } from "node:path"

async function getInputs() {
    const inputs = await readFile(join("..", "inputs.txt"), { encoding: "utf-8" })
    return inputs.split("\n").slice(0, -1)
}

/**
 * @param {Array<string>} inputs
 * @param {function(number, number, number)} calculateFn
 * @returns {number}
*/
function getTotalWrappingPapper(inputs, calculateFn) {
    let totalWrappingPapper = 0
    for (const metric of inputs) {
        const [l, w, h] = metric.split("x")

        totalWrappingPapper += calculateFn(parseInt(l), parseInt(w), parseInt(h))
    }
    return totalWrappingPapper
}

/**
 * @param {number} l
 * @param {number} w
 * @param {number} h
 * @returns {number}
*/
function calculateRectangleWrappingPapper(l, w, h) {
    const baseArea = l * w
    const sideArea = w * h
    const frontArea = l * h

    const wrappingPapper = 2 * baseArea + 2 * sideArea + 2 * frontArea

    return wrappingPapper + Math.min(baseArea, sideArea, frontArea)
}

/**
 * @param {number} l
 * @param {number} w
 * @param {number} h
 * @returns {number}
*/
function calculateRibbon(l, w, h) {
    const basePerimeter = 2 * (l + w)
    const sidePerimeter = 2 * (w + h)
    const frontPerimeter = 2 * (l + h)

    const bow = l * w * h

    return bow + Math.min(basePerimeter, sidePerimeter, frontPerimeter)
}


const inputs = await getInputs()
const totalWrappingPapper = getTotalWrappingPapper(inputs, calculateRectangleWrappingPapper)
const totalRibbon = getTotalWrappingPapper(inputs, calculateRibbon)

console.log(`Total wrapping papper: ${totalWrappingPapper}`)
console.log(`Total ribbon: ${totalRibbon}`)

File: 2015/02/python/main.py
import os
from typing import Callable


def getInputs():
    metrics = None
    with open(path) as inputs:
        metrics = inputs.read()
        metrics = metrics.split("\n")[:-1]
    return metrics


def calculateRectangleWrappingPapper(
        length: int,
        width: int,
        height: int
) -> int:
    baseArea = length * width
    sideArea = width * height
    frontArea = length * height

    wrappingPapper = 2 * baseArea + 2 * sideArea + 2 * frontArea

    return wrappingPapper + min(baseArea, sideArea, frontArea)


def calculateRibbon(
        length: int,
        width: int,
        height: int
) -> int:
    basePerimeter = 2 * (length + width)
    sidePerimeter = 2 * (width + height)
    frontPerimeter = 2 * (length + height)

    bow = length * width * height

    return bow + min(basePerimeter, sidePerimeter, frontPerimeter)


def getTotalPapperSquareFeet(
        metrics: list[str],
        calculateFn: Callable[[int, int, int], int]
) -> int:
    totalPapperSquareFeet = 0

    for metric in metrics:
        metricList = metric.split("x")

        totalPapperSquareFeet += calculateFn(
            int(metricList[0]),
            int(metricList[1]),
            int(metricList[2])
        )

    return totalPapperSquareFeet


if __name__ == "__main__":
    path = os.path.join("..", "inputs.txt")
    metrics = getInputs()
    totalPapperSquareFeet = getTotalPapperSquareFeet(
        metrics,
        calculateRectangleWrappingPapper
    )
    totalRibbonSquareFeet = getTotalPapperSquareFeet(
        metrics,
        calculateRibbon
    )

    print(f'Total wrapping papper: {totalPapperSquareFeet}')
    print(f'Total ribbon: {totalRibbonSquareFeet}')
